### PR TITLE
GG-25963 .NET Thin Client: Handle unsupported features on older server nodes gracefully

### PR DESCRIPTION
**Pull request by ptupitsyn**
* Fix Thin Client version handling bug (could not connect to older nodes)
* Disable Partition Awareness when connected to older servers
* Throw descriptive exception when doing unsupported operations on older server nodes (cluster API)
* Add Thin Client logging (defaults to console)
* Add compatibility tests

### CODE REVIEW

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/ClientFailoverSocket.cs [hunk 1]
```diff
@@ -97,8 +101,15 @@ public ClientFailoverSocket(IgniteClientConfiguration config, Marshaller marsh)
                 throw new IgniteClientException("Failed to resolve all specified hosts.");
             }
 
+            if (_config.Logger == null)
+            {
+                _config.Logger = new ConsoleLogger();
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/ClientFailoverSocket.cs]: This config is a copy of the config from user, and we also copy it in `GetConfiguration` so we can mutate it safely. This instance is not accessible to the user.

I kinda see your concern though, what do you suggest? The idea is to pass this same instance to `ClientSocket`. But at the same time, `ClientSocket` can be instantiated separately. Mutating the config seems like the simplest option.

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/ClientSocket.cs [hunk 2]
```diff
@@ -731,6 +759,26 @@ private void EndRequestsWithError()
                 }
             }
         }
+        
+        /// <summary>
+        /// Validates op code against current protocol version.
+        /// </summary>
+        /// <param name="opId"></param>
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/ClientSocket.cs]: Fixed

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Log/ConsoleLoggerTest.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        [TestCase(LogLevel.Error, LogLevel.Error, true)]
+        [TestCase(LogLevel.Error, LogLevel.Warn, false)]
+        [TestCase(LogLevel.Error, LogLevel.Info, false)]
+        [TestCase(LogLevel.Error, LogLevel.Debug, false)]
+        [TestCase(LogLevel.Error, LogLevel.Trace, false)]
+        [TestCase(LogLevel.Warn, LogLevel.Error, true)]
+        [TestCase(LogLevel.Warn, LogLevel.Warn, true)]
+        [TestCase(LogLevel.Warn, LogLevel.Info, false)]
+        [TestCase(LogLevel.Warn, LogLevel.Trace, false)]
+        public void TestIsEnabledReturnsFalseWhenMinLevelIsGreaterThanLevel(
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/Log/ConsoleLoggerTest.cs]: Not sure I agree here, but if you insist, we can figure out a better name.

File: modules/platforms/dotnet/Apache.Ignite.Core/Log/ConsoleLogger.cs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        public void Log(LogLevel level, string message, object[] args, IFormatProvider formatProvider, string category,
+            string nativeErrorInfo, Exception ex)
+        {
+            if (!IsEnabled(level))
+            {
+                return;
+            }
+
+            var sb = new StringBuilder().AppendFormat(
+                "[{0:HH:mm:ss}] [{1}] [{2}] ", _dateTimeProvider.Now(), level, category);
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core/Log/ConsoleLogger.cs]: It is not, `0:HH:mm:ss` is the format

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/JavaServer/src/main/java/Runner.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        TcpDiscoveryVmIpFinder ipFinder = new TcpDiscoveryVmIpFinder()
+                .setAddresses(Collections.singleton("127.0.0.1:47500..47501"));
+
+        TcpDiscoverySpi discoSpi = new TcpDiscoverySpi().setIpFinder(ipFinder).setSocketTimeout(300);
+
+        IgniteConfiguration cfg = new IgniteConfiguration()
+                .setClientConnectorConfiguration(connectorConfiguration)
+                .setDiscoverySpi(discoSpi);
+
+        Ignition.start(cfg);
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/JavaServer/src/main/java/Runner.java]: We would have to add more logic on C# side:
* Send a char to the process
* Wait for it to shut down
* Does it hang? Kill it then

Killing it straight away is easier, and I don't see any issues with that.

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/JavaServer.cs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    using System.Diagnostics;
+    using System.IO;
+    using System.Text.RegularExpressions;
+    using Apache.Ignite.Core.Client;
+    using Apache.Ignite.Core.Impl.Common;
+    using Apache.Ignite.Core.Impl.Unmanaged;
+    using Apache.Ignite.Core.Tests.Process;
+
+    /// <summary>
+    /// Starts Java server nodes.
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/JavaServer.cs]: Good point, done.

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/ClientServerCompatibilityTest.cs [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+        /// <summary>
+        /// Tests that partition awareness disables automatically on older server versions.
+        /// </summary>
+        [Test]
+        public void TestPartitionAwarenessDisablesAutomaticallyOnVersionsOlderThan140()
+        {
+            using (var client = StartClient())
+            {
+                Assert.IsFalse(client.GetConfiguration().EnablePartitionAwareness);
+                var cache = client.GetOrCreateCache<int, int>(TestContext.CurrentContext.Test.Name);
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/ClientServerCompatibilityTest.cs]: Sorry, I don't get it, can you please elaborate?

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/ClientFailoverSocket.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ashapkin) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/ClientFailoverSocket.cs]: My suggestion is to move this check inside the config class or ignore this.

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/ClientOpExtensionsTest.cs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+                Assert.IsTrue(minVersion >= ClientSocket.Ver100);
+                Assert.IsTrue(minVersion <= ClientSocket.CurrentProtocolVersion);
+            }
+        }
+
+        /// <summary>
+        /// Tests that <see cref="ClientOpExtensions.GetMinVersion"/> returns a specific version for known new features.
+        /// </summary>
+        [Test]
+        public void TestGetMinVersionReturnsSpecificVersionForNewFeatures()
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/ClientOpExtensionsTest.cs]: This test is a unit test for `GetMinVersion` method.
`TestClusterOperationsThrowCorrectExceptionOnVersionsOlderThan150` is an integration test that checks overall behavior. It is good to have both.

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Log/ConsoleLoggerTest.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ashapkin) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/Log/ConsoleLoggerTest.cs]: TestIsEnabled**ReturnsFalse** is wrong, that's it

File: modules/platforms/dotnet/Apache.Ignite.Core/Log/ConsoleLogger.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ashapkin) [reply modules/platforms/dotnet/Apache.Ignite.Core/Log/ConsoleLogger.cs]: Sorry, that was not clear. 
I meant that you need to provide day information alongside with the time: [YYYY-MM-DD:HH:mm:ss]

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/JavaServer.cs [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        }
+
+        /// <summary>
+        /// Updates pom.xml with given Ignite version.
+        /// </summary>
+        private static void ReplaceIgniteVersionInPomFile(string version, string pomFile)
+        {
+            var pomContent = File.ReadAllText(pomFile);
+            pomContent = Regex.Replace(pomContent,
+                @"<version>\d\.\d\.\d</version>",
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/JavaServer.cs]: GridGain switched artifact names when Community Edition was introduced:
* https://gridgainsystems.com/nexus/content/repositories/external/org/gridgain/ignite-core/
* https://gridgainsystems.com/nexus/content/repositories/external/org/apache/ignite/ignite-core/

To use those releases, we would have to use different artifact name depending on version. And donating this PR to Ignite will require more work and cause conflicts later. Too much complexity overall, so let's use Ignite releases.

Good point about regex, fixed.

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/ClientFailoverSocket.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/ClientFailoverSocket.cs]: Config class is not concerned with this. Null logger is a valid config. Let's keep as is then.

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/ClientServerCompatibilityTest.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (ashapkin) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/ClientServerCompatibilityTest.cs]: Well, I personally prefer to divide a multifunctional test (you can think about it when you place multiple non-related assertions) into smaller subcases and give them corresponding names, though it's kinda difficult to follow in real life.

My thoughts:
TestPartitionAwarenessDisablesAutomaticallyOnVersionsOlderThan140 to check config property

TestDisabledPartitionAwarnessPerformsCacheOperations for accessing cache API

File: modules/platforms/dotnet/Apache.Ignite.Core/Log/ConsoleLogger.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core/Log/ConsoleLogger.cs]: If you start Ignite from a NuGet package, you'll get only the time in the default log messages. So I decided to do the same. This is enough for basic logging.

And, as explained in the class docs, for complex logging we expect our users to use NLog/log4net.

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Log/ConsoleLoggerTest.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/Log/ConsoleLoggerTest.cs]: Ohh, I see now :)
Renamed, thanks.

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/ClientFailoverSocket.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/ClientFailoverSocket.cs]: By the way, how do we disable logging in this case? Should user provide some kind of NoOpLogger for this case?

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/ClientServerCompatibilityTest.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/ClientServerCompatibilityTest.cs]: This test actually checks only one thing - that partition awareness is disabled automatically, so cache operations still work.

I agree that more granular tests are better in general, but let's not take it to the extreme of "one assert per test".

File: modules/platforms/dotnet/Apache.Ignite.Core/Log/ConsoleLogger.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ashapkin) [reply modules/platforms/dotnet/Apache.Ignite.Core/Log/ConsoleLogger.cs]: I didn't notice that before. Thanks for the details.

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/ClientFailoverSocket.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/ClientFailoverSocket.cs]: @isapego good point. I propose the following:
* Set `IgniteClientConfiguration.Logger` to `ConsoleLogger` by default, but do this in the `IgniteClientConfiguration` constructor right away, not inside `ClientSocket` during client start
* Users can explicitly set `IgniteClientConfiguration.Logger` to null to disable logging

This should address the above concerns too. What do you think @ashapkin ?

[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/ClientFailoverSocket.cs]: @ptupitsyn sounds good to me. Need to describe this behaviour in configuration documentation though.

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/Cache/ListLogger.cs [hunk 10]
```diff
@@ -21,45 +21,126 @@ namespace Apache.Ignite.Core.Tests.Client.Cache
...
+        /// </summary>
+        public ListLogger(ILogger wrappedLogger = null)
+        {
+            _wrappedLogger = wrappedLogger;
+        }
+
+        /// <summary>
+        /// Gets the entries.
+        /// </summary>
+        public List<Entry> Entries
```
Comment (ashapkin) [inline modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/Cache/ListLogger.cs]: Could we make this getter an Entry class internal?

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/Cache/ListLogger.cs [hunk 11]
```diff
@@ -21,45 +21,126 @@ namespace Apache.Ignite.Core.Tests.Client.Cache
...
+            
+            /** */
+            private readonly string _category;
+
+            /// <summary>
+            /// Initializes a new instance of <see cref="Entry"/> class.
+            /// </summary>
+            public Entry(string message, LogLevel level, string category)
+            {
+                _message = message;
```
Comment (ashapkin) [inline modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/Cache/ListLogger.cs]: Probably a null checking

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/Cache/ListLogger.cs (hunk 10 above)
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/Cache/ListLogger.cs]: Please explain in more detail - how can we move this inside Entry class?

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/Cache/ListLogger.cs (hunk 11 above)
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/Cache/ListLogger.cs]: Added

**Pull request merged**